<!DOCTYPE html> <html lang="en-US"> <head> <meta charset="UTF-8"> <meta http-equiv="X-UA-Compatible" content="IE=Edge"> <title>cgsplice - compgen.io</title> <link rel="shortcut icon" href="/favicon.ico" type="image/x-icon"> <link rel="stylesheet" href="/assets/css/just-the-docs-default.css"> <script async src="https://www.googletagmanager.com/gtag/js?id=UA-83662616-1"></script> <script> window.dataLayer = window.dataLayer || []; function gtag(){dataLayer.push(arguments);} gtag('js', new Date()); gtag('config', 'UA-83662616-1'); </script> <script type="text/javascript" src="/assets/js/just-the-docs.js"></script> <meta name="viewport" content="width=device-width, initial-scale=1"> <!-- Begin Jekyll SEO tag v2.6.1 --> <title>cgsplice | compgen.io</title> <meta name="generator" content="Jekyll v3.8.7" /> <meta property="og:title" content="cgsplice" /> <meta property="og:locale" content="en_US" /> <meta name="description" content="Data wrangling tools for computational genomics" /> <meta property="og:description" content="Data wrangling tools for computational genomics" /> <link rel="canonical" href="https://compgen.io/projects/cgsplice.html" /> <meta property="og:url" content="https://compgen.io/projects/cgsplice.html" /> <meta property="og:site_name" content="compgen.io" /> <script type="application/ld+json"> {"@type":"WebPage","url":"https://compgen.io/projects/cgsplice.html","publisher":{"@type":"Organization","logo":{"@type":"ImageObject","url":"https://compgen.io/assets/img/compgen.png"}},"headline":"cgsplice","description":"Data wrangling tools for computational genomics","@context":"https://schema.org"}</script> <!-- End Jekyll SEO tag --> </head> <body> <svg xmlns="http://www.w3.org/2000/svg" style="display: none;"> <symbol id="link" viewBox="0 0 16 16"> <title>Link</title> <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path> </symbol> </svg> <div class="page-wrap"> <div class="side-bar"> <div class="site-header"> <a href="https://compgen.io/" class="site-title lh-tight"> <div class="site-logo"></div> </a> <button class="menu-button fs-3 js-main-nav-trigger" data-text-toggle="Hide" type="button">Menu</button> </div> <div class="navigation main-nav js-main-nav"> <nav role="navigation" aria-label="Main navigation"> <ul class="navigation-list"><li class="navigation-list-item "><a href="https://compgen.io/" class="navigation-list-link">Home</a></li><li class="navigation-list-item active "><a href="https://compgen.io/projects" class="navigation-list-link">Projects</a><ul class="navigation-list-child-list "><li class="navigation-list-item "><a href="https://compgen.io/ngsutilsj" class="navigation-list-link">ngsutilsj</a> </li><li class="navigation-list-item active "><a href="https://compgen.io/projects/cgsplice.html" class="navigation-list-link active">cgsplice</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/swalign.html" class="navigation-list-link">swalign</a> </li><li class="navigation-list-item nav_gap"><a href="https://compgen.io/cgpipe" class="navigation-list-link">cgpipe</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/sbs" class="navigation-list-link">sbs</a> </li><li class="navigation-list-item nav_gap"><a href="https://compgen.io/tabl" class="navigation-list-link">tabl</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/kdpeaks.html" class="navigation-list-link">kdpeaks</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/compgen-common.html" class="navigation-list-link">compgen-common</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/compgen-cmdline.html" class="navigation-list-link">compgen-cmdline</a> </li><li class="navigation-list-item nav_gap"><a href="https://compgen.io/projects/ngsutils" class="navigation-list-link">NGSUtils</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/tabutils.html" class="navigation-list-link">tabutils</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/basespace-download.html" class="navigation-list-link">basespace-download</a> </li></ul></li><li class="navigation-list-item nav_gap "><a href="https://compgen.io/contact" class="navigation-list-link">Contact / Help</a></li></ul> </nav> </div> <footer class="site-footer"> <p class="text-small text-grey-dk-000 mb-4"> <!-- <a href="https://compgen.io/"><img src="assets/img/compgen-icon.png" style="float:left; width: 32px;"/> compgen.io</a><br/>data wrangling for <br/>computational genomics --> <a href="https://compgen.io/" class="footer_link"><img src="https://compgen.io//assets/img/compgen-slug.png"/></a> </p> </footer> </div> <div class="main-content-wrap js-main-content" tabindex="0"> <div class="main-content"> <div class="page-header js-page-header"> <div class="subtitle"> Data wrangling for computational genomics </div> <ul class="list-style-none text-small aux-nav"> <li class="d-inline-block my-0"><a href="//github.com/compgen-io">GitHub</a></li> </ul> </div> <div class="page"> <nav class="breadcrumb-nav"> <ol class="breadcrumb-nav-list"> <li class="breadcrumb-nav-list-item"><a href="https://compgen.io/projects">Projects</a></li> <li class="breadcrumb-nav-list-item"><span>cgsplice</span></li> </ol> </nav> <div id="main-content" class="page-content" role="main"> <h1 id="cgsplice"> <a href="#cgsplice" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> cgsplice </h1> <p><a href="https://github.com/compgen-io/cgsplice">Github</a></p> <p>cgsplice is a set of tools to support direct measurement of splice junctions in RNA-seq data. cgsplice works by finding and counting reads that span splice junctions. It does this by searching for reads with an "N" gap in their CIGAR alignment string. Next, for all junctions, it determines the donor and acceptor sites and tallies how many reads span each donor/acceptor pair. Finally, using these counts for multiple samples, it will calculate a permuted p-value for how likely it is that a given donor or acceptor had differential splicing between two conditions. Reads supporting a cassette exon, but that don't cross a junction are not included in this model.</p> <h2 id="counting-junction-spanning-reads"> <a href="#counting-junction-spanning-reads" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Counting junction-spanning reads </h2> <p>The <code class="language-plaintext highlighter-rouge">junction-count</code> command will count how many reads span any particular splice junction. Specifically, each junction is split into donor/acceptor pairs. Then for all junctions with a common donor or acceptor, the number of supporting reads is calculated. Only reads that have split-mapping (N gaps in the CIGAR string) are used.</p> <p>Counting is done in a gene model-free method, allowing for discovery of unannotated transcripts (if a suitable aligner is used). In addition to counting reads that span junctions, reads that represent retained introns can also be counted.</p> <h2 id="differential-splicing"> <a href="#differential-splicing" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Differential splicing </h2> <p>The <code class="language-plaintext highlighter-rouge">splice-diff</code> command takes a series of <code class="language-plaintext highlighter-rouge">junction-count</code> outputs and performs a permutation analysis to determine which splicing donor/acceptor sites exhibit differential splicing between two conditions. The statistical metric used is a Student's t-score with Binomial variance. Samples are permuted across different conditions to calculate a p-value for each donor/acceptor.</p> <h2 id="splicing-events"> <a href="#splicing-events" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Splicing events </h2> <p>The <code class="language-plaintext highlighter-rouge">combine-events</code> command merges donor/acceptors into complete splicing events, such as cassette exons. Only donor/acceptor that meet certain criteria (FDR, effect-size) are merged.</p> <h2 id="stats"> <a href="#stats" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Stats </h2> <p>The <code class="language-plaintext highlighter-rouge">bam-stats</code> command can be used to determine how many junction-spanning reads are present in any given sample, among other things. These QC metrics can be used to assess how useful a given BAM file is for junction analysis.</p> <h2 id="installing-cgsplice"> <a href="#installing-cgsplice" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Installing cgsplice </h2> <p>cgsplice is implemented in Java and distributed as a self-executing fast-JAR file. This requires only a recent version of Java to be installed on the system. Otherwise, cgsplice acts as a normal command-line program.</p> </div> </div> </div> </div> </body> </html>
