<!DOCTYPE html> <html lang="en-US"> <head> <meta charset="UTF-8"> <meta http-equiv="X-UA-Compatible" content="IE=Edge"> <title>kdpeaks - compgen.io</title> <link rel="shortcut icon" href="/favicon.ico" type="image/x-icon"> <link rel="stylesheet" href="/assets/css/just-the-docs-default.css"> <script async src="https://www.googletagmanager.com/gtag/js?id=UA-83662616-1"></script> <script> window.dataLayer = window.dataLayer || []; function gtag(){dataLayer.push(arguments);} gtag('js', new Date()); gtag('config', 'UA-83662616-1'); </script> <script type="text/javascript" src="/assets/js/just-the-docs.js"></script> <meta name="viewport" content="width=device-width, initial-scale=1"> <!-- Begin Jekyll SEO tag v2.6.1 --> <title>kdpeaks | compgen.io</title> <meta name="generator" content="Jekyll v3.8.7" /> <meta property="og:title" content="kdpeaks" /> <meta property="og:locale" content="en_US" /> <meta name="description" content="Data wrangling tools for computational genomics" /> <meta property="og:description" content="Data wrangling tools for computational genomics" /> <link rel="canonical" href="https://compgen.io/projects/kdpeaks.html" /> <meta property="og:url" content="https://compgen.io/projects/kdpeaks.html" /> <meta property="og:site_name" content="compgen.io" /> <script type="application/ld+json"> {"@type":"WebPage","url":"https://compgen.io/projects/kdpeaks.html","publisher":{"@type":"Organization","logo":{"@type":"ImageObject","url":"https://compgen.io/assets/img/compgen.png"}},"headline":"kdpeaks","description":"Data wrangling tools for computational genomics","@context":"https://schema.org"}</script> <!-- End Jekyll SEO tag --> </head> <body> <svg xmlns="http://www.w3.org/2000/svg" style="display: none;"> <symbol id="link" viewBox="0 0 16 16"> <title>Link</title> <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path> </symbol> </svg> <div class="page-wrap"> <div class="side-bar"> <div class="site-header"> <a href="https://compgen.io/" class="site-title lh-tight"> <div class="site-logo"></div> </a> <button class="menu-button fs-3 js-main-nav-trigger" data-text-toggle="Hide" type="button">Menu</button> </div> <div class="navigation main-nav js-main-nav"> <nav role="navigation" aria-label="Main navigation"> <ul class="navigation-list"><li class="navigation-list-item "><a href="https://compgen.io/" class="navigation-list-link">Home</a></li><li class="navigation-list-item active "><a href="https://compgen.io/projects" class="navigation-list-link">Projects</a><ul class="navigation-list-child-list "><li class="navigation-list-item "><a href="https://compgen.io/ngsutilsj" class="navigation-list-link">ngsutilsj</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/cgsplice.html" class="navigation-list-link">cgsplice</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/swalign.html" class="navigation-list-link">swalign</a> </li><li class="navigation-list-item nav_gap"><a href="https://compgen.io/cgpipe" class="navigation-list-link">cgpipe</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/sbs" class="navigation-list-link">sbs</a> </li><li class="navigation-list-item nav_gap"><a href="https://compgen.io/tabl" class="navigation-list-link">tabl</a> </li><li class="navigation-list-item active "><a href="https://compgen.io/projects/kdpeaks.html" class="navigation-list-link active">kdpeaks</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/compgen-common.html" class="navigation-list-link">compgen-common</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/compgen-cmdline.html" class="navigation-list-link">compgen-cmdline</a> </li><li class="navigation-list-item nav_gap"><a href="https://compgen.io/projects/ngsutils" class="navigation-list-link">NGSUtils</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/tabutils.html" class="navigation-list-link">tabutils</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/basespace-download.html" class="navigation-list-link">basespace-download</a> </li></ul></li><li class="navigation-list-item nav_gap "><a href="https://compgen.io/contact" class="navigation-list-link">Contact / Help</a></li></ul> </nav> </div> <footer class="site-footer"> <p class="text-small text-grey-dk-000 mb-4"> <!-- <a href="https://compgen.io/"><img src="assets/img/compgen-icon.png" style="float:left; width: 32px;"/> compgen.io</a><br/>data wrangling for <br/>computational genomics --> <a href="https://compgen.io/" class="footer_link"><img src="https://compgen.io//assets/img/compgen-slug.png"/></a> </p> </footer> </div> <div class="main-content-wrap js-main-content" tabindex="0"> <div class="main-content"> <div class="page-header js-page-header"> <div class="subtitle"> Data wrangling for computational genomics </div> <ul class="list-style-none text-small aux-nav"> <li class="d-inline-block my-0"><a href="//github.com/compgen-io">GitHub</a></li> </ul> </div> <div class="page"> <nav class="breadcrumb-nav"> <ol class="breadcrumb-nav-list"> <li class="breadcrumb-nav-list-item"><a href="https://compgen.io/projects">Projects</a></li> <li class="breadcrumb-nav-list-item"><span>kdpeaks</span></li> </ol> </nav> <div id="main-content" class="page-content" role="main"> <h1 id="kdpeaks"> <a href="#kdpeaks" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> kdpeaks </h1> <p><a href="https://github.com/mbreese/kdpeaks">Github</a></p> <p>kdpeaks is an R library for finding peaks in KDE density plots. If you pass the function a kernel density estimate matrix, the function will find all of the peak values in the 2D matrix.</p> <h2 id="installation"> <a href="#installation" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Installation </h2> <p>This package can be installed directly from the Github repository.</p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>library(devtools)
install_github("mbreese/kdpeaks")
</code></pre></div></div> <h2 id="method"> <a href="#method" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Method </h2> <p>kdpeaks works using a variant of an sequencing alignment dynamic programming algorithm.</p> <p>Steps:</p> <ol> <li> <p>For each point in the matrix, find the highest value in all adjacent cells (including itself).</p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code> 1  |  2  |  3
 4  | (5) |  6
 7  |  8  |  9
</code></pre></div> </div> <p>In this case, if we are looking at cell #5, we'd determine the maximum value between all nine of these cells. This step is repeated for all cells in the matrix. This generated a "flow matrix" indicating the direction of the maxima.</p> </li> <li> <p>Peaks are identified as the cells where that are their own maxima. In the above example, if 5 is the highest peak of the adjacent cells, then 5 is it's own maxima. Each position in the matrix is then assigned to a peak based on the flow. Peaks that are adjacent to other peaks (within a set radius) are merged.</p> </li> <li> <p>Peaks are sorted in their order of magnitude (their height).</p> </li> <li> <p>The top peaks are calculated based on the mean and standard deviation of the peak heights.</p> </li> </ol> <p>Data from all peaks are returned from the function, sorted by their height.</p> <h2 id="example"> <a href="#example" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Example </h2> <p>Here is an example KDE plot for a log2-ratio vs BAF data in whole-genome sequencing.</p> <p><img src="/assets/img/chr1_plot2.png" alt="KDE plot" /></p> <p>kdpeaks was used to find the peaks as identified by the 7 points. 1-D density plots are shown for each axis as a comparison.</p> <p>Here is how this plot was generated.</p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>k &lt;- kde2d(chr1.df$log2ratio, chr1.df$baf, n = 100)

peaks&lt;-find_peaks(k, combine_within = 5)

peaks.df &lt;- data.frame(x=peaks$x[1:peaks$top_peaks], y=peaks$y[1:peaks$top_peaks], n=as.factor(peaks$num[1:peaks$top_peaks]))
</code></pre></div></div> <p>The x/y coordinates of all peaks is returned as <code class="language-plaintext highlighter-rouge">peaks$x</code> and <code class="language-plaintext highlighter-rouge">peaks$y</code>. The number of top peaks is returned in <code class="language-plaintext highlighter-rouge">peaks$top_peaks</code> (given the z-score, 1.65 by default). The height of the peaks is given in <code class="language-plaintext highlighter-rouge">peaks$z</code>. The area that each peak contains is given in <code class="language-plaintext highlighter-rouge">peak$size</code>. The flow matrix is returned as <code class="language-plaintext highlighter-rouge">peak$peak.m</code>.</p> </div> </div> </div> </div> </body> </html>
