<!DOCTYPE html> <html lang="en-US"> <head> <meta charset="UTF-8"> <meta http-equiv="X-UA-Compatible" content="IE=Edge"> <title>ngsutilsj - compgen.io</title> <link rel="shortcut icon" href="/favicon.ico" type="image/x-icon"> <link rel="stylesheet" href="/assets/css/just-the-docs-default.css"> <script async src="https://www.googletagmanager.com/gtag/js?id=UA-83662616-1"></script> <script> window.dataLayer = window.dataLayer || []; function gtag(){dataLayer.push(arguments);} gtag('js', new Date()); gtag('config', 'UA-83662616-1'); </script> <script type="text/javascript" src="/assets/js/just-the-docs.js"></script> <meta name="viewport" content="width=device-width, initial-scale=1"> <!-- Begin Jekyll SEO tag v2.6.1 --> <title>ngsutilsj | compgen.io</title> <meta name="generator" content="Jekyll v3.8.7" /> <meta property="og:title" content="ngsutilsj" /> <meta property="og:locale" content="en_US" /> <meta name="description" content="Data wrangling tools for computational genomics" /> <meta property="og:description" content="Data wrangling tools for computational genomics" /> <link rel="canonical" href="https://compgen.io/ngsutilsj" /> <meta property="og:url" content="https://compgen.io/ngsutilsj" /> <meta property="og:site_name" content="compgen.io" /> <script type="application/ld+json"> {"@type":"WebPage","url":"https://compgen.io/ngsutilsj","publisher":{"@type":"Organization","logo":{"@type":"ImageObject","url":"https://compgen.io/assets/img/compgen.png"}},"headline":"ngsutilsj","description":"Data wrangling tools for computational genomics","@context":"https://schema.org"}</script> <!-- End Jekyll SEO tag --> </head> <body> <svg xmlns="http://www.w3.org/2000/svg" style="display: none;"> <symbol id="link" viewBox="0 0 16 16"> <title>Link</title> <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path> </symbol> </svg> <div class="page-wrap"> <div class="side-bar"> <div class="site-header"> <a href="https://compgen.io/" class="site-title lh-tight"> <div class="site-logo"></div> </a> <button class="menu-button fs-3 js-main-nav-trigger" data-text-toggle="Hide" type="button">Menu</button> </div> <div class="navigation main-nav js-main-nav"> <nav role="navigation" aria-label="Main navigation"> <ul class="navigation-list"><li class="navigation-list-item "><a href="https://compgen.io/" class="navigation-list-link">Home</a></li><li class="navigation-list-item active "><a href="https://compgen.io/projects" class="navigation-list-link">Projects</a><ul class="navigation-list-child-list "><li class="navigation-list-item active "><a href="https://compgen.io/ngsutilsj" class="navigation-list-link active">ngsutilsj</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/cgsplice.html" class="navigation-list-link">cgsplice</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/swalign.html" class="navigation-list-link">swalign</a> </li><li class="navigation-list-item nav_gap"><a href="https://compgen.io/cgpipe" class="navigation-list-link">cgpipe</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/sbs" class="navigation-list-link">sbs</a> </li><li class="navigation-list-item nav_gap"><a href="https://compgen.io/tabl" class="navigation-list-link">tabl</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/kdpeaks.html" class="navigation-list-link">kdpeaks</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/compgen-common.html" class="navigation-list-link">compgen-common</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/compgen-cmdline.html" class="navigation-list-link">compgen-cmdline</a> </li><li class="navigation-list-item nav_gap"><a href="https://compgen.io/projects/ngsutils" class="navigation-list-link">NGSUtils</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/tabutils.html" class="navigation-list-link">tabutils</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/basespace-download.html" class="navigation-list-link">basespace-download</a> </li></ul></li><li class="navigation-list-item nav_gap "><a href="https://compgen.io/contact" class="navigation-list-link">Contact / Help</a></li></ul> </nav> </div> <footer class="site-footer"> <p class="text-small text-grey-dk-000 mb-4"> <!-- <a href="https://compgen.io/"><img src="assets/img/compgen-icon.png" style="float:left; width: 32px;"/> compgen.io</a><br/>data wrangling for <br/>computational genomics --> <a href="https://compgen.io/" class="footer_link"><img src="https://compgen.io//assets/img/compgen-slug.png"/></a> </p> </footer> </div> <div class="main-content-wrap js-main-content" tabindex="0"> <div class="main-content"> <div class="page-header js-page-header"> <div class="subtitle"> Data wrangling for computational genomics </div> <ul class="list-style-none text-small aux-nav"> <li class="d-inline-block my-0"><a href="//github.com/compgen-io">GitHub</a></li> </ul> </div> <div class="page"> <nav class="breadcrumb-nav"> <ol class="breadcrumb-nav-list"> <li class="breadcrumb-nav-list-item"><a href="https://compgen.io/projects">Projects</a></li> <li class="breadcrumb-nav-list-item"><span>ngsutilsj</span></li> </ol> </nav> <div id="main-content" class="page-content" role="main"> <h1 id="ngsutilsj"> <a href="#ngsutilsj" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> ngsutilsj </h1> <p><a href="https://github.com/compgen-io/ngsutilsj">Github</a></p> <p>ngsutilsj is an updated java port of the NGSUtils toolkit. This new version is largely a Java port of the the most commonly used tools from NGSUtils, with some additions thrown in. It is also a library, with utility classes for use in other various NGS related software (such as <a href="/cgsplice">cgsplice</a>).</p> <p>Java was chosen for the ease of installation and relative speed (in comparison to the Python NGSUtils). The processing speed for gzipped compressed files was a major reason for the new update. This version has also been optimized for working on high-memory HPC clusters and streaming data analysis (to minimize disk IO).</p> <h2 id="installation"> <a href="#installation" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Installation </h2> <p>ngsutilsj is distributed as a self-executing fat-JAR file. This means that for installation, all one needs is a working copy of Java and the <code class="language-plaintext highlighter-rouge">ngsutilsj</code> file. Unlike other JAR-file based NGS packages, ngsutilsj includes a shell script shim to make it executable like a traditional Unix program. This means that to install the program, you need to copy the <code class="language-plaintext highlighter-rouge">ngsutilsj</code> file to somewhere in your <code class="language-plaintext highlighter-rouge">$PATH</code>.</p> <h2 id="commands-available"> <a href="#commands-available" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Commands available </h2> <p>There are many commands available. For information on an individual command, run <code class="language-plaintext highlighter-rouge">ngsutilsj help command</code></p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ngsutilsj - Data wrangling for NGS
---------------------------------------

Usage: ngsutilsj cmd [options]

Available commands:
[bam]
  bam-basecall     - For a BAM file, output the basecalls (ACGTN) at each genomic position.
  bam-best         - With reads mapped to two bam references, determine which reads mapped best to each
  bam-bins         - Quickly count the number of reads that fall into bins (bins assigned based on 5' end of the first read)
  bam-check        - Checks a BAM file to make sure it is valid
  bam-concat       - Concatenates BAM files (handles @RG, @PG)
  bam-count        - Counts the number of reads for genes (GTF), within a BED region, or by bins (--gtf, --bed, or --bins required)
  bam-coverage*    - Scans an aligned BAM file and calculates the number of reads covering each base
  bam-discord      - Extract all discordant reads from a BAM file
  bam-dups         - Flags or removes duplicate reads
  bam-expressed    - For a BAM file, output all regions with significant coverage in BED format.
  bam-filter       - Filters out reads based upon various criteria
  bam-readgroup    - Add a read group ID to each read in a BAM file
  bam-sample*      - Create a whitelist of read names sampled randomly from a file
  bam-split        - Split a BAM file into smaller files
  bam-stats        - Stats about a BAM file and the library orientation
  bam-tobed*       - Writes read positions to a BED6 file
  bam-tobedgraph   - Calculate coverave for an aligned BAM file in BedGraph format.
  bam-tofastq      - Export the read sequences from a BAM file in FASTQ format

[bed]
  bed-clean        - Cleans score entries to be an integer
  bed-count        - Given reference and query BED files, count the number of query regions that are contained within each reference region
  bed-nearest      - Given reference and query BED files, for each query region, find the nearest reference region
  bed-reduce       - Merge overlaping BED regions
  bed-resize       - Resize BED regions (extend or shrink)
  bed-tobed3       - Convert a BED3+ file to a strict BED3 file
  bed-tobed6       - Convert a BED6+ file to a strict BED6 file
  bed-tofasta      - Extract FASTA sequences based on BED coordinates

[fasta]
  fasta-filter     - Filter out sequences from a FASTA file
  fasta-gc         - Determine the GC% for a given region or bins
  fasta-genreads*  - Generate mock reads from a reference FASTA file
  fasta-mask       - Mask regions of a FASTA reference
  fasta-names      - Display sequence names from a FASTA file
  fasta-split      - Split a FASTA file into a new file for each sequence present
  fasta-subseq     - Extract subsequences from a FASTA file (optionally, indexed)
  fasta-tag        - Add prefix/suffix to FASTA sequence names
  fasta-wrap       - Change the sequence wrapping length of a FASTA file

[fastq]
  fastq-barcode    - Given Illumina 1.8+ naming, find the lane/barcodes included
  fastq-check      - Verify a FASTQ single, paired, or interleaved file(s)
  fastq-demux      - Splits a FASTQ file based on lane/barcode values
  fastq-filter     - Filters reads from a FASTQ file.
  fastq-merge      - Merges two FASTQ files (R1/R2) into one interleaved file.
  fastq-separate   - Splits an interleaved FASTQ file by read number.
  fastq-sort       - Sorts a FASTQ file
  fastq-split      - Splits an FASTQ file into smaller files
  fastq-stats      - Statistics about a FASTQ file
  fastq-tobam      - Converts a FASTQ file (or two paired files) into an unmapped BAM file
  fastq-tofasta    - Convert FASTQ sequences to FASTA format

[gtf]
  gtf-export       - Export gene annotations from a GTF file as BED regions
  gtf-geneinfo*    - Calculate information about genes (based on GTF model)

[annotation]
  annotate-gtf     - Annotate GTF gene regions (for tab-delimited text, BED, or BAM input)
  annotate-repeat* - Calculates Repeat masker annotations

[vcf]
  vcf-annotate     - Annotate a VCF file
  vcf-chrfix       - Changes the reference (chrom) format (Ensembl/UCSC)
  vcf-export       - Export information from a VCF file
  vcf-filter       - Filter a VCF file
  vcf-tobed        - Export allele positions from a VCF file to BED format

[help]
  help             - Help for a specific command
  license          - Show the license

* = experimental command
</code></pre></div></div> </div> </div> </div> </div> </body> </html>
