<!DOCTYPE html> <html lang="en-US"> <head> <meta charset="UTF-8"> <meta http-equiv="X-UA-Compatible" content="IE=Edge"> <title>Tutorial: FASTQ Pre-processing - compgen.io</title> <link rel="shortcut icon" href="/favicon.ico" type="image/x-icon"> <link rel="stylesheet" href="/assets/css/just-the-docs-default.css"> <script async src="https://www.googletagmanager.com/gtag/js?id=UA-83662616-1"></script> <script> window.dataLayer = window.dataLayer || []; function gtag(){dataLayer.push(arguments);} gtag('js', new Date()); gtag('config', 'UA-83662616-1'); </script> <script type="text/javascript" src="/assets/js/just-the-docs.js"></script> <meta name="viewport" content="width=device-width, initial-scale=1"> <!-- Begin Jekyll SEO tag v2.6.1 --> <title>Tutorial: FASTQ Pre-processing | compgen.io</title> <meta name="generator" content="Jekyll v3.8.7" /> <meta property="og:title" content="Tutorial: FASTQ Pre-processing" /> <meta property="og:locale" content="en_US" /> <meta name="description" content="Data wrangling tools for computational genomics" /> <meta property="og:description" content="Data wrangling tools for computational genomics" /> <link rel="canonical" href="https://compgen.io/tutorials/fastq-preprocessing" /> <meta property="og:url" content="https://compgen.io/tutorials/fastq-preprocessing" /> <meta property="og:site_name" content="compgen.io" /> <script type="application/ld+json"> {"@type":"WebPage","url":"https://compgen.io/tutorials/fastq-preprocessing","publisher":{"@type":"Organization","logo":{"@type":"ImageObject","url":"https://compgen.io/assets/img/compgen.png"}},"headline":"Tutorial: FASTQ Pre-processing","description":"Data wrangling tools for computational genomics","@context":"https://schema.org"}</script> <!-- End Jekyll SEO tag --> </head> <body> <svg xmlns="http://www.w3.org/2000/svg" style="display: none;"> <symbol id="link" viewBox="0 0 16 16"> <title>Link</title> <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path> </symbol> </svg> <div class="page-wrap"> <div class="side-bar"> <div class="site-header"> <a href="https://compgen.io/" class="site-title lh-tight"> <div class="site-logo"></div> </a> <button class="menu-button fs-3 js-main-nav-trigger" data-text-toggle="Hide" type="button">Menu</button> </div> <div class="navigation main-nav js-main-nav"> <nav role="navigation" aria-label="Main navigation"> <ul class="navigation-list"><li class="navigation-list-item "><a href="https://compgen.io/" class="navigation-list-link">Home</a></li><li class="navigation-list-item "><a href="https://compgen.io/projects" class="navigation-list-link">Projects</a><ul class="navigation-list-child-list "><li class="navigation-list-item "><a href="https://compgen.io/ngsutilsj" class="navigation-list-link">ngsutilsj</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/cgsplice.html" class="navigation-list-link">cgsplice</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/swalign.html" class="navigation-list-link">swalign</a> </li><li class="navigation-list-item nav_gap"><a href="https://compgen.io/cgpipe" class="navigation-list-link">cgpipe</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/sbs" class="navigation-list-link">sbs</a> </li><li class="navigation-list-item nav_gap"><a href="https://compgen.io/tabl" class="navigation-list-link">tabl</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/kdpeaks.html" class="navigation-list-link">kdpeaks</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/compgen-common.html" class="navigation-list-link">compgen-common</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/compgen-cmdline.html" class="navigation-list-link">compgen-cmdline</a> </li><li class="navigation-list-item nav_gap"><a href="https://compgen.io/projects/ngsutils" class="navigation-list-link">NGSUtils</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/tabutils.html" class="navigation-list-link">tabutils</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/basespace-download.html" class="navigation-list-link">basespace-download</a> </li></ul></li><li class="navigation-list-item nav_gap "><a href="https://compgen.io/contact" class="navigation-list-link">Contact / Help</a></li></ul> </nav> </div> <footer class="site-footer"> <p class="text-small text-grey-dk-000 mb-4"> <!-- <a href="https://compgen.io/"><img src="assets/img/compgen-icon.png" style="float:left; width: 32px;"/> compgen.io</a><br/>data wrangling for <br/>computational genomics --> <a href="https://compgen.io/" class="footer_link"><img src="https://compgen.io//assets/img/compgen-slug.png"/></a> </p> </footer> </div> <div class="main-content-wrap js-main-content" tabindex="0"> <div class="main-content"> <div class="page-header js-page-header"> <div class="subtitle"> Data wrangling for computational genomics </div> <ul class="list-style-none text-small aux-nav"> <li class="d-inline-block my-0"><a href="//github.com/compgen-io">GitHub</a></li> </ul> </div> <div class="page"> <nav class="breadcrumb-nav"> <ol class="breadcrumb-nav-list"> <li class="breadcrumb-nav-list-item"><a href="">Tutorials</a></li> <li class="breadcrumb-nav-list-item"><span>Tutorial: FASTQ Pre-processing</span></li> </ol> </nav> <div id="main-content" class="page-content" role="main"> <h1 class="no_toc" id="tutorial-fastq-pre-processing"> <a href="#tutorial-fastq-pre-processing" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Tutorial: FASTQ Pre-processing </h1> <p>The first step in NGS analysis is pre-processing your <a href="/tutorials/formats#fastq">FASTQ</a> files. The following tutorial uses <a href="/ngsutilsj/fastq-filter">ngsutilsj fastq-filter</a> to process and filter FASTQ files.</p> <h2 class="no_toc text-delta" id="overview"> <a href="#overview" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Overview </h2> <ol id="markdown-toc"> <li><a href="#sequence-qc-tools" id="markdown-toc-sequence-qc-tools">Sequence QC tools</a> <ol> <li><a href="#fastqc" id="markdown-toc-fastqc">FastQC</a></li> <li><a href="#ngsutilsj-fastq-stats" id="markdown-toc-ngsutilsj-fastq-stats">ngsutilsj fastq-stats</a></li> </ol> </li> <li><a href="#dna-seq-filtering-methods" id="markdown-toc-dna-seq-filtering-methods">DNA-seq filtering methods</a> <ol> <li><a href="#examples" id="markdown-toc-examples">Examples</a></li> </ol> </li> <li><a href="#rna-seq-filtering-methods" id="markdown-toc-rna-seq-filtering-methods">RNA-seq filtering methods</a></li> </ol> <h2 id="sequence-qc-tools"> <a href="#sequence-qc-tools" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Sequence QC tools </h2> <p>It's important to measure certain quality control metrics with FASTQ files. You should run the same QC checks against the files before and after pre- processing. This will help you assess the quality of your data and measure how well the pre-processing is working. QC monitoring is a critical part of NGS analysis that is often neglected. Without monitoring read QC, you may inadvertently use low-quality data, which could produce invalid analysis. (and a lot of wasted effort!)</p> <h3 id="fastqc"> <a href="#fastqc" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> FastQC </h3> <p>One commonly used program to assess the quality of FASTQ data is <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>. FastQC is a good tool for assessing the initial quality of FASTQ reads and provides a number of quality control metrics, including per-base sequence quality, GC %, per-base call frequency, over-represented sequences, and adapter content. The final report from FastQC can be saves as a ZIP or HTML report for use in pipelines. However, FastQC can take a while to run on larger files.</p> <h3 id="ngsutilsj-fastq-stats"> <a href="#ngsutilsj-fastq-stats" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> ngsutilsj fastq-stats </h3> <p>For high-throughput streaming workflows, where the FASTQ data is processed on-the-fly, it may not be possible (or practical) to save a copy of the data. In this case, you need a tool that can calculate QC metrics for FASTQ data that is streamed as part of a stdin/stdout pipeline. <a href="/ngsutilsj/fastq-stats">ngsutilsj fastq-stats</a> supports this method and calculates per-base sequence quality, GC %, per-base call frequencgy, and adapter content. Additionally, <a href="/ngsutilsj/fastq-stats">ngsutilsj fastq-stats</a> also support interleaved FASTQ files and reporting first/second read stats separately. The final report from <a href="/ngsutilsj/fastq-stats">ngsutilsj fastq-stats</a> is a raw text file that requires further processing to produce figures.</p> <p>Importantly, when inserted into a FASTQ pipeline, <a href="/ngsutilsj/fastq-stats">ngsutilsj fastq-stats</a> requires no extra processing time, and minimal memory/CPU resources.</p> <h2 id="dna-seq-filtering-methods"> <a href="#dna-seq-filtering-methods" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> DNA-seq filtering methods </h2> <p>Pre-processing DNA-seq files is more straightforward than RNA-seq, so we'll start there. For DNA-seq, the primary concern is removing low-quality bases and reads before sending them through the computationally intensive alignment process. The more that we clean up the data before alignment, the better the aligned data will be.</p> <p>Here are some of the ways to cleanup FASTQ reads with <a href="/ngsutilsj/fastq-filter">ngsutilsj fastq-filter</a>:</p> <ol> <li> <p>Remove reads based on a name include/exclude list If you have a specific list of reads to keep or drop, they can be passed as a <code class="language-plaintext highlighter-rouge">--include</code> or <code class="language-plaintext highlighter-rouge">--exclude</code> respectively.</p> </li> <li> <p>Fixed base count 5'/3' trim<br /> You can remove a fixed number of bases from the 5' or 3' ends of a read with the <code class="language-plaintext highlighter-rouge">--prefixtrim</code> or <code class="language-plaintext highlighter-rouge">--suffixtrim</code> options.</p> </li> <li> <p>Remove low quality bases from the 5' and 3' ends<br /> It is common for there to be stretches of low-quality base calls at the 3' end of a read. It is less common to see low-quality calls at the 5' end. You can set a fixed threshold for minimum call quality at the 5' or 3' ends using the <code class="language-plaintext highlighter-rouge">--prefixqual</code> and <code class="language-plaintext highlighter-rouge">--suffixqual</code> options. For example, if the threshold is set at 3, then any base with a Phred quality score less than 3 will be removed from the ends of the reads.</p> </li> <li> <p>Trim adapter sequences<br /> If the fragment being sequenced is shorter than the read length, you will see adapter sequences as part of the read sequence. Because these are artifacts of the sequencing reaction and not part of the organism, it is important to remove these sequences. There are three options for removing adapters: <code class="language-plaintext highlighter-rouge">--trim-seq</code>, <code class="language-plaintext highlighter-rouge">--trim-seq1</code>, <code class="language-plaintext highlighter-rouge">--trim-seq2</code>. The first option trims an adapter from all sequences in a file. The second and third options trim the adapter from the first and second reads, respectively (assuming an interleaved FASTQ file). If you have a non-interleaved file, then <code class="language-plaintext highlighter-rouge">--trim-seq</code> and <code class="language-plaintext highlighter-rouge">--trim-seq1</code> are equivalent. As extra options for adapter trimming: <code class="language-plaintext highlighter-rouge">--trim-pct</code> sets the required match percentage a the read has to have to match the adapter. The default is 0.9, or 9 out of 10 bases must match. The second option is <code class="language-plaintext highlighter-rouge">--trim-overlap</code>, which sets the minimum amount of sequence must match at the 3' end of the read to trigger trimming. The default value is 6 bases.<br /> <br /> Note: adapter trimming is done using a sliding window, so it is necessarily computationally intensive. It should only be done when you really need to do it.<br /> <br /> For Illumina sequencing, the recommended adapter trimming sequences can be set using the <code class="language-plaintext highlighter-rouge">--trim-illumina</code> option. These sequences are read-specific and are shown below:</p> <ul> <li>read1: AGATCGGAAGAGCACACGTC</li> <li>read2: AGATCGGAAGAGCGTCGTGT</li> </ul> </li> <li> <p>Limit abiguous/wildcard (N) bases<br /> Reads with too many <em>N</em> calls can be removed using the <code class="language-plaintext highlighter-rouge">--wildcard</code> option.</p> </li> <li> <p>Size selection After all of the above trimming, the final read filter can require a certain minimum read length with the <code class="language-plaintext highlighter-rouge">--size</code> option.</p> </li> <li> <p>Required valid paired reads (requires interleaved FASTQ file) If you have an interleaved FASTQ file, you can use the <code class="language-plaintext highlighter-rouge">--paired</code> option to require that both reads from a fragment need to pass the filters. Otherwise, you can end up with an unbalanced FASTQ file, which can lead to unpredictable results in the downstream analysis. However, this option is only available when using an interleaved FASTQ file. Processing two separate files and trying to merge them after the fact is computationally difficult and may require a lot of memory to resolve mismatches between the files.</p> </li> </ol> <p>Note: filters are applied in the above order, so that (for example) low-quality bases at the 5'/3' ends can be trimmed away before triggering the ambiguous base call filters.</p> <h3 id="examples"> <a href="#examples" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Examples </h3> <p>Here is one way to filter a single-end FASTQ file, or a split paired-end (not interleaved) file:</p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code># trim 3' low-quality bases
# remove reads with more than 2 N's
# require the final size to be at least 50bp

ngsutilsj fastq-filter --size 50 --wildcard 2 --suffixqual 3 input.fastq.gz &gt; filtered.fastq
</code></pre></div></div> <p><br /> Here is a more complete set of filters, including a merging step to combine paired-end files into one interleaved FASTQ file first.</p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code># trim 3' low-quality bases
# remove illumina adapters
# remove reads with more than 2 N's
# require the final size to be at least 50bp
# require reads to be properly paired

ngsutilsj fastq-merge input_R1.fastq.gz input_R2.fastq.gz | \
ngsutilsj fastq-filter --size 50 --wildcard 2 --suffixqual 3 --trim-illumina --paired - | \
gzip &gt; filtered.fastq.gz
</code></pre></div></div> <p><br /> Finally, we can add <a href="/ngsutilsj/fastq-stats">ngsutilsj fastq-stats</a> into the mix to calculate some summary statistics on the fly for both the raw FASTQ reads and the post-filtering reads. By merging the paired FASTQ files and using <a href="/ngsutilsj/fastq-stats">ngsutilsj fastq-stats</a>, we can efficiently pre-process the FASTQ reads in one pipeline and limit unnecessary disk I/O.</p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ngsutilsj fastq-merge input_R1.fastq.gz input_R2.fastq.gz | \
ngsutilsj fastq-stats --pipe -o raw.stats.txt - | \
ngsutilsj fastq-filter --size 50 --wildcard 2 --suffixqual 3 --trim-illumina --paired - | \
ngsutilsj fastq-stats --pipe -o filtered.stats.txt - | \
gzip &gt; filtered.fastq.gz
</code></pre></div></div> <p><br /> As shown above and following <em>nix conventions, many of the <a href="/ngsutilsj">ngsutilsj</a> tools can operate in streaming mode by specifying "<code class="language-plaintext highlighter-rouge">-</code>" as the input filename, reading input from *stdin</em>. <a href="/ngsutilsj/fastq-stats">ngsutilsj fastq-stats</a> can be coerced into passing the incoming FASTQ data through to <em>stdout</em> with the <code class="language-plaintext highlighter-rouge">--pipe</code> option.</p> <h2 id="rna-seq-filtering-methods"> <a href="#rna-seq-filtering-methods" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> RNA-seq filtering methods </h2> <p>Pre-processing RNA-seq data can be done in the same manner as DNA-seq data, with the same general options. But, with RNA-seq data, there is another filtering step that you should consider: ribosomal sequence filtering. A good library prep will remove the majority of rRNA sequences before they are sequenced. If this depletion fails, then many of reads will be from rRNAs. This makes tracking the abundance of rRNAs an important QC metric to ensure consistent data.</p> <p>Filtering out rRNA reads is covered in-depth in the <a href="/tutorials/rnaseq-star">RNA-Seq workflow</a> tutorial.</p> </div> </div> </div> </div> </body> </html>
