<!DOCTYPE html> <html lang="en-US"> <head> <meta charset="UTF-8"> <meta http-equiv="X-UA-Compatible" content="IE=Edge"> <title>Tutorial: File formats - compgen.io</title> <link rel="shortcut icon" href="/favicon.ico" type="image/x-icon"> <link rel="stylesheet" href="/assets/css/just-the-docs-default.css"> <script async src="https://www.googletagmanager.com/gtag/js?id=UA-83662616-1"></script> <script> window.dataLayer = window.dataLayer || []; function gtag(){dataLayer.push(arguments);} gtag('js', new Date()); gtag('config', 'UA-83662616-1'); </script> <script type="text/javascript" src="/assets/js/just-the-docs.js"></script> <meta name="viewport" content="width=device-width, initial-scale=1"> <!-- Begin Jekyll SEO tag v2.6.1 --> <title>Tutorial: File formats | compgen.io</title> <meta name="generator" content="Jekyll v3.8.7" /> <meta property="og:title" content="Tutorial: File formats" /> <meta property="og:locale" content="en_US" /> <meta name="description" content="Data wrangling tools for computational genomics" /> <meta property="og:description" content="Data wrangling tools for computational genomics" /> <link rel="canonical" href="https://compgen.io/tutorials/formats" /> <meta property="og:url" content="https://compgen.io/tutorials/formats" /> <meta property="og:site_name" content="compgen.io" /> <script type="application/ld+json"> {"@type":"WebPage","url":"https://compgen.io/tutorials/formats","publisher":{"@type":"Organization","logo":{"@type":"ImageObject","url":"https://compgen.io/assets/img/compgen.png"}},"headline":"Tutorial: File formats","description":"Data wrangling tools for computational genomics","@context":"https://schema.org"}</script> <!-- End Jekyll SEO tag --> </head> <body> <svg xmlns="http://www.w3.org/2000/svg" style="display: none;"> <symbol id="link" viewBox="0 0 16 16"> <title>Link</title> <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path> </symbol> </svg> <div class="page-wrap"> <div class="side-bar"> <div class="site-header"> <a href="https://compgen.io/" class="site-title lh-tight"> <div class="site-logo"></div> </a> <button class="menu-button fs-3 js-main-nav-trigger" data-text-toggle="Hide" type="button">Menu</button> </div> <div class="navigation main-nav js-main-nav"> <nav role="navigation" aria-label="Main navigation"> <ul class="navigation-list"><li class="navigation-list-item "><a href="https://compgen.io/" class="navigation-list-link">Home</a></li><li class="navigation-list-item "><a href="https://compgen.io/projects" class="navigation-list-link">Projects</a><ul class="navigation-list-child-list "><li class="navigation-list-item "><a href="https://compgen.io/ngsutilsj" class="navigation-list-link">ngsutilsj</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/cgsplice.html" class="navigation-list-link">cgsplice</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/swalign.html" class="navigation-list-link">swalign</a> </li><li class="navigation-list-item nav_gap"><a href="https://compgen.io/cgpipe" class="navigation-list-link">cgpipe</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/sbs" class="navigation-list-link">sbs</a> </li><li class="navigation-list-item nav_gap"><a href="https://compgen.io/tabl" class="navigation-list-link">tabl</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/kdpeaks.html" class="navigation-list-link">kdpeaks</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/compgen-common.html" class="navigation-list-link">compgen-common</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/compgen-cmdline.html" class="navigation-list-link">compgen-cmdline</a> </li><li class="navigation-list-item nav_gap"><a href="https://compgen.io/projects/ngsutils" class="navigation-list-link">NGSUtils</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/tabutils.html" class="navigation-list-link">tabutils</a> </li><li class="navigation-list-item "><a href="https://compgen.io/projects/basespace-download.html" class="navigation-list-link">basespace-download</a> </li></ul></li><li class="navigation-list-item nav_gap "><a href="https://compgen.io/contact" class="navigation-list-link">Contact / Help</a></li></ul> </nav> </div> <footer class="site-footer"> <p class="text-small text-grey-dk-000 mb-4"> <!-- <a href="https://compgen.io/"><img src="assets/img/compgen-icon.png" style="float:left; width: 32px;"/> compgen.io</a><br/>data wrangling for <br/>computational genomics --> <a href="https://compgen.io/" class="footer_link"><img src="https://compgen.io//assets/img/compgen-slug.png"/></a> </p> </footer> </div> <div class="main-content-wrap js-main-content" tabindex="0"> <div class="main-content"> <div class="page-header js-page-header"> <div class="subtitle"> Data wrangling for computational genomics </div> <ul class="list-style-none text-small aux-nav"> <li class="d-inline-block my-0"><a href="//github.com/compgen-io">GitHub</a></li> </ul> </div> <div class="page"> <nav class="breadcrumb-nav"> <ol class="breadcrumb-nav-list"> <li class="breadcrumb-nav-list-item"><a href="">Tutorials</a></li> <li class="breadcrumb-nav-list-item"><span>Tutorial: File formats</span></li> </ol> </nav> <div id="main-content" class="page-content" role="main"> <h1 class="no_toc" id="tutorial-file-formats"> <a href="#tutorial-file-formats" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Tutorial: File formats </h1> <h2 class="no_toc text-delta" id="overview"> <a href="#overview" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Overview </h2> <ol id="markdown-toc"> <li><a href="#sequencing-formats" id="markdown-toc-sequencing-formats">Sequencing formats</a> <ol> <li><a href="#fasta" id="markdown-toc-fasta">FASTA</a> <ol> <li><a href="#variations" id="markdown-toc-variations">Variations</a></li> <li><a href="#example" id="markdown-toc-example">Example</a></li> <li><a href="#tools" id="markdown-toc-tools">Tools</a></li> <li><a href="#external-links" id="markdown-toc-external-links">External links</a></li> </ol> </li> <li><a href="#fastq" id="markdown-toc-fastq">FASTQ</a> <ol> <li><a href="#variations-1" id="markdown-toc-variations-1">Variations</a></li> <li><a href="#example-1" id="markdown-toc-example-1">Example</a></li> <li><a href="#external-links-1" id="markdown-toc-external-links-1">External links</a></li> </ol> </li> <li><a href="#unaligned-bam-file" id="markdown-toc-unaligned-bam-file">Unaligned BAM file</a></li> <li><a href="#ngsutilsj-support" id="markdown-toc-ngsutilsj-support">ngsutilsj support</a></li> </ol> </li> </ol> <h2 id="sequencing-formats"> <a href="#sequencing-formats" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Sequencing formats </h2> <h3 id="fasta"> <a href="#fasta" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> FASTA </h3> <p>FASTA files are the de-facto standard for storing and sharing DNA/RNA/amino acid sequences. It is a text-file format with a header line followed by one or more lines of sequence. The header line starts with a '&gt;' character, which may be used for parsing the file. The remainder of the header line consists of a name, optionally followed by a comment. The name and comment are separated by whitespace. The name must not contain a whitespace (<code class="language-plaintext highlighter-rouge">Â </code>, or <code class="language-plaintext highlighter-rouge">\t</code>) character.</p> <p>The remaining lines consist of DNA, RNA, or amino acid sequences using IUPAC notation. There are no technical specifications for line length or wrapping, but it is common practice for long sequences (genes/genomes) to be have a consistent line length. However, it is not required for the sequence to have any breaks. Blank lines and any whitespace in the sequence line(s) should be ignored.</p> <p>FASTA is the most common format for distributing reference genome sequences. Reference genomes will contain one sequence for each chromosome. In addition to chromosomes, some reference genomes will also include unplaced contigs, which are sequences known to be part of the organism's genome, but the exact position is unclear.</p> <h4 id="variations"> <a href="#variations" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Variations </h4> <p>FASTA files are commonly stored as <code class="language-plaintext highlighter-rouge">gzip</code> compressed files, due to their large size and general space inefficiency. However, files that need to be randomly accessed need to be stored uncompressed, or compressed with a tool like <code class="language-plaintext highlighter-rouge">bgzip</code> (although the latter is less common).</p> <h4 id="example"> <a href="#example" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Example </h4> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&gt;sequence1 This is the comment line
ATCGATCGATCGACTACGACTACGACGACGACATCGACATCTACT
GGCGCGCGCTAGAGCTAGCTTGAGATAAATCGACTAGCGACTGAG
CTATCTTCTCTATATATTTAAAAAGCGCAACTACTGACTA
&gt;seq2
AATAGCGCGCGCGCGCTCATATATCTATATATAAAAACCTACTAC
GACTACGACTATCGATCGATTATCGGTATCGTATCGGTATTATTA
TTTAATGCGCGCGCGCCGACTAGCTAGCTATCGATCGATCGATCG
ACTACGACTACGACGACGACATCGACATCTACT
</code></pre></div></div> <h4 id="tools"> <a href="#tools" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Tools </h4> <p>FASTA files can be indexed and queried with the program <code class="language-plaintext highlighter-rouge">samtools faidx</code>. This requires a well-formatted FASTA file with consistent line lengths.</p> <p><a href="/ngsutilsj">ngsutilsj</a> contains a number of tools for managing FASTA files, including indexed FASTA files. These tools including tagging sequence names, masking regions of sequences, splitting a FASTA file by sequence name, changing the line wrapping for a file, or generating mock <a href="/tutorials/formats#fastq">FASTQ</a> reads.</p> <h4 id="external-links"> <a href="#external-links" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> External links </h4> <p><a href="https://en.wikipedia.org/wiki/FASTA_format">https://en.wikipedia.org/wiki/FASTA_format</a></p> <h3 id="fastq"> <a href="#fastq" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> FASTQ </h3> <p>FASTQ is the most common sequencing read format. FASTQ files are text files with a four-line record for each sequence. The first line starts with an '@' character and the unique name for the read. The second line is the sequence. The third line starts with the '+' character, and may optionally contain the read name again. The fourth line is the quality score for each of the basecalls in <a href="https://en.wikipedia.org/wiki/Phred_quality_score">Phred</a> scale.</p> <h4 id="variations-1"> <a href="#variations-1" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Variations </h4> <p>The sequence and quality lines in the FASTQ record may be word-wrapped, like FASTA files, but this is not seen very often.</p> <p>FASTQ files are almost always stored compressed, usually with <code class="language-plaintext highlighter-rouge">gzip</code>, although <code class="language-plaintext highlighter-rouge">bzip2</code> or <code class="language-plaintext highlighter-rouge">xz</code> compression can also be used. FASTQ record information can also be stored in other compressed formats, such as <a href="/tutorials/formats#unalignedbam">unaligned BAM</a> files. However, almost all publicly available data is distributed as gzip-compressed FASTQ files.</p> <p>Paired-end sequencing data is commonly stored as two separate FASTQ files, one for each read. But, it is also possible to store both reads in the same file. These are called <em>interleaved</em> FASTQ files. Many aligners support interleaved files out of the box for paired end data. For those that don't an adapter tool can be used to convert interleaved files to non-interleaved files using named FIFO pipes. Interleaved FASTQ files are slightly have slightly more efficient compression ratios when compared to using two separate FASTQ files, but the main benefit is the need to only manage one file per sample. Some filtering steps can be more efficient using interleaved files and some filtering steps with the <a href="/ngsutilsj/fastq-filter">ngsutilsj fastq-filter</a> tool require interleaved files (e.g. <code class="language-plaintext highlighter-rouge">--paired</code>).</p> <h4 id="example-1"> <a href="#example-1" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Example </h4> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>@seq1 This is the comment line
ATCGATCGATCGACTACGACTACGACGACGACATCGACATCTACT
+
BBBB,(,,7AA,&lt;((,A&lt;,,,FKAF,,,F,F7F,7,,,AA##@!@
</code></pre></div></div> <h4 id="external-links-1"> <a href="#external-links-1" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> External links </h4> <p>https://en.wikipedia.org/wiki/FASTQ_format</p> <h3 id="unaligned-bam-file"> <a href="#unaligned-bam-file" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Unaligned BAM file </h3> <p>BAM files are typically associated with read alignments to a genome, but they can also be used to store unaligned/raw sequences too. Unaligned BAM files store the same read information as a FASTQ file (name, sequence, and quality scores) and they can also store multiple reads in the same file (like interleaved FASTQ files). Also, unaligned BAM files are compressed, and store quality score information in an optimized manner versus character encoding. Even with the extra overhead of the BAM format, these factors make this a slightly more efficient way to store raw sequencing reads over FASTQ files.</p> <h3 id="ngsutilsj-support"> <a href="#ngsutilsj-support" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> ngsutilsj support </h3> <p>Both FASTQ (compressed or not) and unaligned BAM files can be used as a source for sequencing reads with <a href="/ngsutilsj">ngsutilsj</a>. Compression status is determined automatically by ngsutilsj.</p> </div> </div> </div> </div> </body> </html>
