<!DOCTYPE html>
<html class="no-js" lang="en-US" prefix="og: http://ogp.me/ns# fb: http://ogp.me/ns/fb#">
<head>
    <meta charset="utf-8">

        <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-83662616-1', 'auto');
  ga('send', 'pageview');

</script>


    <base href="https://compgen.io/">
    <title>compgen.io - ngsutilsj</title>
    <link rel="canonical" href="https://compgen.io/ngsutilsj/">

    <link href='https://fonts.googleapis.com/css?family=Lato:400,300,100' rel='stylesheet' type='text/css'>

<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.6/css/bootstrap.min.css" integrity="sha384-1q8mTJOASx8j1Au+a5WDVnPi2lkFfwwEAa8hDDdjZlpLegxhjVME1fgjWPGmkzs7" crossorigin="anonymous">
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css" crossorigin="anonymous">





<link rel="stylesheet/less" type="text/css" href="/css/style.less" />

<script src="/js/less.min.js" type="text/javascript"></script>
<link rel="shortcut icon" href="/favicon.ico" />

</head>
<body lang="en">

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.6/js/bootstrap.min.js" integrity="sha384-0mSbJDEHialfmuBBQP6A4Qrprq5OVfW37PRR3j5ELqxss1yVqOtnepnHVP9aJ7xS" crossorigin="anonymous"></script>

<div class="header_wrap">
<div class="header">
  <div class="logo"><a href="/"><img src="/img/compgen-logo.png" alt="compgen.io"/></a></div>
  <div class="slogan">data wrangling for<br/>computational genomics</div>
  <ul class="links">
      <li class="links dropdown">
          <a href="/projects" style="padding-right: 0;"><i class="fa fa-cogs"></i>&nbsp;Projects</a>
          <a class="dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false" style="padding-left: 2px;" id="dropdownMenu1">
            <span class="fa fa-sort-down"></span>
            <span class="sr-only">Toggle Projects Dropdown</span>
          </a>
          <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
              <li class="dropdown-header">NGS data analysis</li>
            <li><a href="/ngsutilsj">ngsutilsj (newer)</a></li>
            <li><a href="/ngsutils">NGSUtils (older)</a></li>
            <li><a href="/cgsplice">cgsplice</a></li>
            <li><a href="/sqz">SQZ</a></li>
            <li role="separator" class="divider"></li>
            <li class="dropdown-header">Data analysis pipelines</li>
            <li><a href="/cgpipe">cgpipe</a></li>
            <li><a href="/sbs">SBS</a></li>
            <li role="separator" class="divider"></li>
            <li class="dropdown-header">Misc. tools</li>
            <li><a href="/basespace-download">Basespace-download</a></li>
            
            <li><a href="/swalign">swalign</a></li>
            <li><a href="/tabutils">tabutils</a></li>
            <li role="separator" class="divider"></li>
            <li class="dropdown-header">Programming libraries</li>
            <li><a href="/compgen-common">compgen-common</a></li>
            <li><a href="/compgen-cmdline">compgen-cmdline</a></li>
          </ul>
      </li>
      <li class="links"><a href="/tutorials"><i class="fa fa-book"></i>&nbsp;Tutorials</a></li>

      <li class="links"><a href="/citations"><i class="fa fa-external-link-squaref"></i> Citations</a></li>
      <li class="links"><a href="/contact"><i class="fa fa-question-circle"></i>&nbsp;Contact</a></li>


  </ul>
  <div class="clearing">&nbsp;</div>
</div>
</div>

<div class="project_bar_wrap">
  <ul class="project_bar">
    <li class="heading"><a href="/projects">Projects</a> > ngsutilsj</li>
    <li class="project_link"><a href="/ngsutilsj" class="active">Overview</a></li>
  
  
    <li class="project_link"><a href="https://github.com/compgen-io/ngsutilsj"><i class="fa fa-github"></i> GitHub project</a></li>
  
    <li class="project_link"><a href="https://github.com/compgen-io/ngsutilsj/issues"><i class="fa fa-bug"></i> Issues</a></li>


  
    <li class="project_link"><a href="https://github.com/compgen-io/ngsutilsj/releases"><i class="fa fa-download"></i> Download</a></li>


  </ul>
</div>
<div class="main_wrap">
  <section id="main">
    <h1 id="title"><i class="fa fa-cogs"></i> ngsutilsj</h1>
    <div>
          <article id="content" class="tutorial">
             

<p>ngsutilsj is an updated java port of the NGSUtils toolkit. This new version is largely a Java port of the
the most commonly used tools from NGSUtils, with some additions thrown in. It is also a library, with utility
classes for use in other various NGS related software (such as <a href="/cgsplice">cgsplice</a>).</p>

<p>Java was chosen for the ease of installation and relative speed (in comparison to the Python NGSUtils).
The processing speed for gzipped compressed files was a major reason for the new update. This version has
also been optimized for working on high-memory HPC clusters and streaming data analysis (to minimize disk IO).</p>

<h2 id="installation">Installation</h2>

<p>ngsutilsj is distributed as a self-executing fat-JAR file. This means that for installation, all one needs is
a working copy of Java and the <code>ngsutilsj</code> file. Unlike other JAR-file based NGS packages, ngsutilsj includes a
shell script shim to make it executable like a traditional Unix program. This means that to install the program,
you need to copy the <code>ngsutilsj</code> file to somewhere in your <code>$PATH</code>.</p>

<h2 id="commands-available">Commands available</h2>

<p>There are many commands available. For information on an individual command, run <code>ngsutilsj help command</code></p>

<pre><code>ngsutilsj - Data wrangling for NGS
---------------------------------------

Usage: ngsutilsj cmd [options]

Available commands:
[bam]
  bam-basecall     - For a BAM file, output the basecalls (ACGTN) at each genomic position.
  bam-best         - With reads mapped to two bam references, determine which reads mapped best to each
  bam-bins         - Quickly count the number of reads that fall into bins (bins assigned based on 5' end of the first read)
  bam-check        - Checks a BAM file to make sure it is valid
  bam-concat       - Concatenates BAM files (handles @RG, @PG)
  bam-count        - Counts the number of reads for genes (GTF), within a BED region, or by bins (--gtf, --bed, or --bins required)
  bam-coverage*    - Scans an aligned BAM file and calculates the number of reads covering each base
  bam-discord      - Extract all discordant reads from a BAM file
  bam-dups         - Flags or removes duplicate reads
  bam-expressed    - For a BAM file, output all regions with significant coverage in BED format.
  bam-filter       - Filters out reads based upon various criteria
  bam-readgroup    - Add a read group ID to each read in a BAM file
  bam-sample*      - Create a whitelist of read names sampled randomly from a file
  bam-split        - Split a BAM file into smaller files
  bam-stats        - Stats about a BAM file and the library orientation
  bam-tobed*       - Writes read positions to a BED6 file
  bam-tobedgraph   - Calculate coverave for an aligned BAM file in BedGraph format.
  bam-tofastq      - Export the read sequences from a BAM file in FASTQ format

[bed]
  bed-clean        - Cleans score entries to be an integer
  bed-count        - Given reference and query BED files, count the number of query regions that are contained within each reference region
  bed-nearest      - Given reference and query BED files, for each query region, find the nearest reference region
  bed-reduce       - Merge overlaping BED regions
  bed-resize       - Resize BED regions (extend or shrink)
  bed-tobed3       - Convert a BED3+ file to a strict BED3 file
  bed-tobed6       - Convert a BED6+ file to a strict BED6 file
  bed-tofasta      - Extract FASTA sequences based on BED coordinates

[fasta]
  fasta-filter     - Filter out sequences from a FASTA file
  fasta-gc         - Determine the GC% for a given region or bins
  fasta-genreads*  - Generate mock reads from a reference FASTA file
  fasta-mask       - Mask regions of a FASTA reference
  fasta-names      - Display sequence names from a FASTA file
  fasta-split      - Split a FASTA file into a new file for each sequence present
  fasta-subseq     - Extract subsequences from a FASTA file (optionally, indexed)
  fasta-tag        - Add prefix/suffix to FASTA sequence names
  fasta-wrap       - Change the sequence wrapping length of a FASTA file

[fastq]
  fastq-barcode    - Given Illumina 1.8+ naming, find the lane/barcodes included
  fastq-check      - Verify a FASTQ single, paired, or interleaved file(s)
  fastq-demux      - Splits a FASTQ file based on lane/barcode values
  fastq-filter     - Filters reads from a FASTQ file.
  fastq-merge      - Merges two FASTQ files (R1/R2) into one interleaved file.
  fastq-separate   - Splits an interleaved FASTQ file by read number.
  fastq-sort       - Sorts a FASTQ file
  fastq-split      - Splits an FASTQ file into smaller files
  fastq-stats      - Statistics about a FASTQ file
  fastq-tobam      - Converts a FASTQ file (or two paired files) into an unmapped BAM file
  fastq-tofasta    - Convert FASTQ sequences to FASTA format

[gtf]
  gtf-export       - Export gene annotations from a GTF file as BED regions
  gtf-geneinfo*    - Calculate information about genes (based on GTF model)

[annotation]
  annotate-gtf     - Annotate GTF gene regions (for tab-delimited text, BED, or BAM input)
  annotate-repeat* - Calculates Repeat masker annotations

[vcf]
  vcf-annotate     - Annotate a VCF file
  vcf-chrfix       - Changes the reference (chrom) format (Ensembl/UCSC)
  vcf-export       - Export information from a VCF file
  vcf-filter       - Filter a VCF file
  vcf-tobed        - Export allele positions from a VCF file to BED format

[help]
  help             - Help for a specific command
  license          - Show the license

* = experimental command
</code></pre>

          </article>
    </div>
  </section>
</div>

<div class="footer_wrap">
<div class="footer">
<div>

<b>compgen.io</b>&nbsp;&nbsp;&middot;&nbsp;&nbsp; Copyright &copy;2016-2018
<br/><br/>

github: <a href="https://github.com/compgen-io">compgen-io</a> 
<a href="https://github.com/mbreese">mbreese</a> 
<a href="https://github.com/ngsutils">ngsutils</a>
</div>
</div>
</div>

</body>
